File: backend/internal/delivery/aftersales.go
package delivery

import (
	"net/http"
	"strconv"

	"github.com/labstack/echo/v4"
	"github.com/typefunco/dealer_dev_platform/internal/utils"
)

// AfterSalesDealerResponse представляет дилера с данными After Sales для API.
type AfterSalesDealerResponse struct {
	ID                    string  `json:"id"`
	Name                  string  `json:"name"`
	City                  string  `json:"city"`
	RStockPercent         *int    `json:"rStockPercent"`         // Recommended Stock
	WStockPercent         *int    `json:"wStockPercent"`         // Warranty Stock
	FlhPercent            *int    `json:"flhPercent"`            // Foton Labor Hours
	FlhSharePercent       *string `json:"flhSharePercent"`       // Foton Labour Hours Share
	WarrantyHours         *int    `json:"warrantyHours"`         // Foton Warranty Hours
	ServiceContractsHours *int    `json:"serviceContractsHours"` // Service Contracts
	AsTrainings           *bool   `json:"asTrainings"`           // AS Trainings status
	CSI                   *string `json:"csi"`                   // Customer Satisfaction Index
	AsDecision            *string `json:"asDecision"`            // AS Decision
	SparePartsSalesQ3     *string `json:"sparePartsSalesQ3"`     // Spare Parts Sales Q3
	SparePartsSalesYtd    *string `json:"sparePartsSalesYtd"`    // Spare Parts Sales YTD %
}

// GetAfterSalesData возвращает данные послепродажного обслуживания с поддержкой фильтров.
// @Summary Get After Sales data
// @Description Получение данных послепродажного обслуживания с фильтрацией по региону, году, кварталу и дилерам
// @Tags aftersales
// @Accept json
// @Produce json
// @Param region query string false "Region filter"
// @Param quarter query string false "Quarter filter (Q1, Q2, Q3, Q4)"
// @Param year query int false "Year filter"
// @Param dealer_ids query string false "Comma-separated dealer IDs"
// @Param limit query int false "Limit for pagination"
// @Param offset query int false "Offset for pagination"
// @Param sort_by query string false "Sort field"
// @Param sort_order query string false "Sort order (asc, desc)"
// @Success 200 {array} AfterSalesDealerResponse
// @Failure 400 {object} ErrorResponse
// @Failure 500 {object} ErrorResponse
// @Router /api/aftersales [get]
func (s *Server) GetAfterSalesData(c echo.Context) error {
	// Парсим параметры фильтрации с помощью универсальной функции
	filters := utils.ParseFilterParamsFromContext(c)

	// Устанавливаем значения по умолчанию для обратной совместимости
	defaults := map[string]interface{}{
		"region":  "all-russia",
		"quarter": "Q1",
		"year":    2024,
	}
	utils.SetDefaultFilters(filters, defaults)

	// Получение данных из сервиса с фильтрами
	afterSalesList, err := s.afterSalesService.GetAfterSalesWithFilters(c.Request().Context(), filters)
	if err != nil {
		s.logger.Error("GetAfterSalesData: failed to get after sales data",
			"filters", filters,
			"error", err,
		)
		return c.JSON(http.StatusInternalServerError, ErrorResponse{
			Error: "Failed to get after sales data",
		})
	}

	// Преобразование в API response
	response := make([]AfterSalesDealerResponse, 0, len(afterSalesList))
	for _, as := range afterSalesList {
		response = append(response, AfterSalesDealerResponse{
			ID:                    strconv.FormatInt(int64(as.DealerID), 10),
			Name:                  as.DealerNameRu,
			City:                  as.City,
			RStockPercent:         &as.RecommendedStock,
			WStockPercent:         &as.WarrantyStock,
			FlhPercent:            &as.FotonLaborHours,
			WarrantyHours:         &as.FotonWarrantyHours,
			ServiceContractsHours: &as.ServiceContracts,
			AsTrainings:           &as.ASTrainings,
			CSI:                   as.CSI,
			AsDecision:            &as.ASDecision,
		})
	}

	return c.JSON(http.StatusOK, response)
}

File: frontend/src/api/aftersales.ts
// API клиент для работы с данными After Sales
// Интеграция с backend After Sales API

const API_BASE_URL = import.meta.env.DEV ? '/api' : 'http://localhost:8080/api'

export interface AfterSalesDealer {
  id: string;
  name: string;
  city: string;
  rStockPercent: number;
  wStockPercent: number;
  flhPercent: number;
  flhSharePercent: string;
  warrantyHours: number;
  serviceContractsHours: number;
  asTrainings: boolean;
  csi: string;
  asDecision: 'Needs development' | 'Planned Result' | 'Find New Candidate' | 'Close Down';
  sparePartsSalesQ3: string;
  sparePartsSalesYtd: string;
}

export interface AfterSalesResponse {
  dealers: AfterSalesDealer[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

export interface AfterSalesFilters {
  search?: string;
  region?: string;
  decision?: string;
  page?: number;
  limit?: number;
}

/**
 * Получить данные After Sales дилеров
 */
export async function getAfterSalesData(filters?: AfterSalesFilters): Promise<AfterSalesResponse> {
  const params = new URLSearchParams();
  
  if (filters?.search) params.append('search', filters.search);
  if (filters?.region) params.append('region', filters.region);
  if (filters?.decision) params.append('decision', filters.decision);
  if (filters?.page) params.append('page', filters.page.toString());
  if (filters?.limit) params.append('limit', filters.limit.toString());
  
  // Добавляем обязательные параметры quarter и year
  params.append('quarter', 'Q1');
  params.append('year', '2024');

  const response = await fetch(`${API_BASE_URL}/aftersales?${params.toString()}`, {
    credentials: 'include', // Включаем cookies для аутентификации
  });
  
  if (!response.ok) {
    throw new Error(`Failed to fetch after sales data: ${response.statusText}`);
  }
  
  const dealers = await response.json();
  
  // Преобразуем ответ бэкенда в формат, ожидаемый фронтендом
  return {
    dealers: Array.isArray(dealers) ? dealers : [],
    pagination: {
      page: filters?.page || 1,
      limit: filters?.limit || 10,
      total: Array.isArray(dealers) ? dealers.length : 0,
      totalPages: Math.ceil((Array.isArray(dealers) ? dealers.length : 0) / (filters?.limit || 10))
    }
  };
}

// Хелперы

/**
 * Список доступных решений по After Sales
 */
export const AFTER_SALES_DECISIONS = [
  'Planned Result',
  'Needs development',
  'Find New Candidate',
  'Close Down'
] as const;

export type AfterSalesDecision = typeof AFTER_SALES_DECISIONS[number];
